src/diagnostics/diagnostic.jl

```diff
@@ -0,0 +1,694 @@
+
+# When the reduction is nothing, do nothing
+reset_accumulator!(_, reduction_time_func::Nothing) = nothing
+
+# If we have a reduction, we have to reset the accumulator to its neutral state. (If we
+# don't have a reduction, we don't have to do anything)
+#
+# ClimaAtmos defines methods for identity_of_reduction for standard reduction_time_func in
+# reduction_identities.jl
+function reset_accumulator!(diag_accumulator, reduction_time_func)
+    # identity_of_reduction works by dispatching over operation
+    identity = identity_of_reduction(reduction_time_func)
+    float_type = eltype(diag_accumulator)
+    identity_ft = convert(float_type, identity)
+    diag_accumulator .= identity_ft
+end
+
+# When the reduction is nothing, we do not need to accumulate anything
+accumulate!(_, _, reduction_time_func::Nothing) = nothing
+
+# When we have a reduction, apply it between the accumulated value one
+function accumulate!(diag_accumulator, diag_storage, reduction_time_func)
+    diag_accumulator .= reduction_time_func.(diag_accumulator, diag_storage)
+end
+
+"""
+    get_callbacks_from_diagnostics(diagnostics, storage, counters)
+
+
+Translate a list of diagnostics into a list of callbacks.
+
+Positional arguments
+=====================
+
+- `diagnostics`: List of `ScheduledDiagnosticIterations` that have to be converted to
+                 callbacks. We want to have `ScheduledDiagnosticIterations` here so that we
+                 can define callbacks that occur at the end of every N integration steps.
+
+- `storage`: Dictionary that maps a given `ScheduledDiagnosticIterations` to a potentially
+             pre-allocated area of memory where to save the newly computed results.
+
+- `accumulator`: Dictionary that maps a given `ScheduledDiagnosticIterations` to a potentially
+                 pre-allocated area of memory where to accumulate results.
+
+- `counters`: Dictionary that maps a given `ScheduledDiagnosticIterations` to the counter
+              that tracks how many times the given diagnostics was computed from the last
+              time it was output to disk.
+
+"""
+function get_callbacks_from_diagnostics(
+    diagnostics,
+    storage,
+    accumulators,
+    counters,
+)
+    # We have two types of callbacks: to compute and accumulate diagnostics, and to dump
+    # them to disk. Note that our callbacks do not contain any branching
+
+    # storage is used to pre-allocate memory and to accumulate partial results for those
+    # diagnostics that perform reductions.
+
+    callbacks = Any[]
+
+    for diag in diagnostics
+        variable = diag.variable
+
+        compute_callback =
+            integrator -> begin
+                # FIXME: Change when ClimaCore overrides .= for us to avoid multiple allocations
+                storage[diag] .= variable.compute!(
+                    storage[diag],
+                    integrator.u,
+                    integrator.p,
+                    integrator.t,
+                )
```

Comment:
```suggestion
                variable.compute!(
                    storage[diag],
                    integrator.u,
                    integrator.p,
                    integrator.t,
                )
```

or if you really want to be pedantic
```
                storage[diag] = variable.compute!(
                    storage[diag],
                    integrator.u,
                    integrator.p,
                    integrator.t,
                )
```